lib/src/outputs/advanced_file_output.dart

```diff
@@ -0,0 +1,147 @@
+    this.directory,
+    this.file,
+    this.overrideExisting = false,
+    this.encoding = utf8,
+    List<Level>? writeImmediately,
+    this.maxDelay = const Duration(seconds: 2),
+    this.maxBufferSize = 2000,
+    this.maxLogFileSizeMB = 1,
+  })  : writeImmediately = writeImmediately ??
+            [
+              Level.error,
+              Level.fatal,
+              Level.warning,
+              // ignore: deprecated_member_use_from_same_package
+              Level.wtf,
+            ],
+        assert(
+          (file != null ? 1 : 0) + (directory != null ? 1 : 0) == 1,
+          'Either file or directory must be set',
+        );
+
+  final File? file;
+  final Directory? directory;
+
+  final bool overrideExisting;
+  final Encoding encoding;
+
+  final List<Level> writeImmediately;
+  final Duration maxDelay;
+  final int maxLogFileSizeMB;
+  final int maxBufferSize;
+
+  IOSink? _sink;
+  File? _targetFile;
+  Timer? _bufferWriteTimer;
+  Timer? _targetFileUpdater;
+
+  final List<OutputEvent> _buffer = [];
+
+  bool get dynamicFilesMode => directory != null;
+  File? get targetFile => _targetFile;
+
+  @override
+  Future<void> init() async {
+    if (dynamicFilesMode) {
+      //we use sync directory check to avoid losing
+      //potential initial boot logs in early crash scenarios
+      if (!directory!.existsSync()) {
+        directory!.createSync(recursive: true);
+      }
+
+      _targetFileUpdater = Timer.periodic(
+        const Duration(minutes: 1),
+        (_) => _updateTargetFile(),
+      );
+    }
+
+    _bufferWriteTimer = Timer.periodic(maxDelay, (_) => _writeOutBuffer());
+    await _updateTargetFile(); //run first setup without waiting for timer tick
+  }
+
+  @override
+  void output(OutputEvent event) {
+    _buffer.add(event);
+    // If event level is present in writeImmediately, write out the buffer
+    // along with any other possible elements that accumulated in it since
+    // the last timer tick
+    // Also write out if buffer is overfilled
+    if (_buffer.length > maxBufferSize ||
+        writeImmediately.contains(event.level)) {
+      _writeOutBuffer();
+    }
+  }
+
+  void _writeOutBuffer() {
+    if (_sink == null) return; //wait until _sink becomes available
+    for (final event in _buffer) {
+      _sink?.writeAll(event.lines, Platform.lineTerminator);
+      _sink?.writeln();
+    }
+    _buffer.clear();
+  }
+
+  Future<void> _updateTargetFile() async {
+    if (!dynamicFilesMode) {
+      await _openFile(file!);
+      return;
+    }
+
+    final t = DateTime.now();
+    final newName =
+        '${t.year}-${t.month.twoDigits}-${t.day.twoDigits}_${t.hour.twoDigits}-${t.minute.twoDigits}-${t.second.twoDigits}-${t.millisecond.threeDigits}';
+    if (_targetFile == null) {
+      // just create a new file on first boot
+      await _openFile(File('${directory!.path}/${newName}_init.txt'));
+    } else {
+      final proposed = File('${directory!.path}/${newName}_next.txt');
```

Comment:
I pushed a commit with this change, but IMO it's questionable.
If logger wasn't destroyed properly, the latest.log file may not be renamed to the actual timestamp, and then, on the next launch, we'll have to detect it and rename to something like 1234-56...log.lost to avoid overwriting it with a new file. IMO, it's better to avoid renaming and stick with a format where the name is fixed right from the beginning.

Reply:
Why not just check if a `latest.log` file exists on startup?
Then you can either reuse it (append) or rename it now if it is greater than max size?
No need to handle any crashes exceptionally.

You had two different file names even before my comment, `init` and `next`, I just tried to make them more familiar for developers.

Reply:
So you mean that latest.log file should always be appended until it reaches the size limit, even if the program has rebooted many times, it still has to write everything to the same file with the size limit being the only factor that pushes a new file and renames the older one, right?

Reply:
Yes, I assumed that was your basic idea. It also closely reassembles something known as a RollingFile handler, which usually behaves similar to this.
Would you like to change something about that approach?

Addendum: You could also add other constraints for the rotation of log files, however, this would significantly increase the complexity of this class and is probably better implemented in a separate output.

Reply:
Thanks for the changes! I am fine with the new rotating logic, so I guess we can resolve this now.
How about you?

Reply:
Yep, everything seems alright